### DISCUSSION TITLE
Email Invitations

### DISCUSSION THREAD

**Discussion by philippneugebauer** [ckulka/baikal-docker#61] [Ideas]
Have you thought about solutions for enabling email invitations?

When I try that, I get the following error:
```
sh: 1: /usr/sbin/sendmail: not found
```

**Comment by ckulka** [comment]
I figure installing sendmail is all that's needed (https://stackoverflow.com/questions/38299776/usr-sbin-sendmail-not-found/38299891). I'll post an update when I updated the Dockerfile, would you be able to test it then?

**Comment by philippneugebauer** [comment]
yeah, for sure. I tried it and got stuck with the postfix installation :(

**Comment by ckulka** [comment]
Ok, it's building now - based on previous Docker Hub builds, that'll take about 1.5h. Can you then please give it a try with one of the following (well, ideally both if you can) images?

* `ckulka/baikal:experimental-apache`
* `ckulka/baikal:experimental-nginx`

**Comment by philippneugebauer** [comment]
Yeah, I found that SO answer, too. I am still not sure, if that's the right thing to do since that means to really host such an smtp server which comes with quite some effort. I think it would be nicer to be able to include a service or at least having it configured outside the baikal container?

**Comment by philippneugebauer** [comment]
that description might be interesting for the setup: https://mchodled.com/posts/2020/Apr/03/de-google-your-calendar-with-baikal/#configure-email-invitation-plugin

**Comment by philippneugebauer** [comment]
I will provide a PR with my changes soon. It's working but I struggle to set the values in `/etc/msmtprc` with ENVs

**Comment by ahgraber** [comment]
@philippneugebauer I'm working on this myself.  

Using the solution from the mchodled.com link, I think it's possible to mount the files as volumes, so the end user should just have an `msmtprc` or `msmtp_php` locally with their config, and it is mounted in the docker-compose:

```yaml
#...
services:
  baikal:
    image: ckulka/baikal:nginx
    container_name: baikal
    restart: unless-stopped
    volumes:
      - ./config:/var/www/baikal/config
      - ./data:/var/www/baikal/Specific
      - ./msmtp_php:/etc/.msmtp_php
```

The real trick is to update the `sendmail_path` directly, and I think you'd have to rely on some bash scripting in the entrypoint/cmd script`start.sh`.  You could do something like:

```bash
sed -i 's/;sendmail_path =/sendmail_path = "/usr/bin/msmtp -C /etc/.msmtp_php --logfile /var/log/msmtp.log -a ${SENDMAIL_ACCOUNT} -t"/' /etc/php/8.0/fpm/php.ini
```
where SENDMAIL_ACCOUNT is an environmental variable

I haven't had time to test yet, though.

**Comment by philippneugebauer** [comment]
have you had a look at my PR? @ahgraber

**Comment by ahgraber** [comment]
Apologies, was moving fast trying to solve the same problem and just saw the comment about struggling to set values -- didn't see the PR link.  Wasn't trying to be pedantic, just wanted to help :)

Your solution is specific to the :apache version?

**Comment by philippneugebauer** [comment]
I did not mean it like that, I rather wanted to point out the PR to get some feedback from you

I just haven't implemented it for nginx since I wanted to get some feedback first to avoid duplicate work

**Comment by ahgraber** [comment]
I'll follow up with comments on the PR itself.

**Comment by lennvilardi** [comment]
Is this problem closed ? Is there a tutorial to use email invites ?

**Comment by philippneugebauer** [comment]
there is a possible option in https://github.com/ckulka/baikal-docker/pull/32

**Comment by MrPinrel** [comment]
I recently installed this container (great work!) on my Synology NAS and would love to get the email invites working. How would I go about enabling this? Do I need to download the experimental nginx version? Or is it going to get included in the latest release soon?

Happy to be a tester but would need some help on how to get started. Also, I use a mail relay, not sure if this makes a difference?

**Comment by philippneugebauer** [comment]
build the dockerfile from #32 on your own and use it

**Comment by MrPinrel** [comment]
> build the dockerfile from #32 on your own and use it

Thanks, I got it to build but it is complaining that msmtprc is not installed in the synology box. I found msmtprc and tried to install it, but it tells me that there is no suitable C compiler found to do that....now trying to figure out how to install that. Let me know if you have any ideas.

For a broader distribution it might be good to include those steps in the automated installation.

**Comment by MrPinrel** [comment]
update, I'm not having any luck getting msmtprc to build in my synology box. I get errors about libraries missing and errors in the MAKE file. Does anybody know where I could find the already built images? It is a DS918+ with the Intel Celeron processor.

**Comment by lennvilardi** [comment]
Seriously considering installing Nextcloud just to get email notification, it's too complicated to compile msmtp for a DS918+... is there another way (simplier?) ?

**Comment by MrPinrel** [comment]
> Seriously considering installing Nextcloud just to get email notification, it's too complicated to compile msmtp for a DS918+... is there another way (simplier?) ?

Lenn thanks for this note, it caused me to look into installing nextcloud to a caldav server with working invites running on DSM 7.0. I followed[ these instructions](https://mariushosting.com/how-to-install-nextcloud-on-your-synology-nas/) and it was really easy. First I tried to install in Docker but then I decided to do the native install.

**Comment by vaskozl** [comment]
I've been a happy radicale user for a few years now, after initially migrating away from baikal a couple of years ago. I've now setup baikal again, the reason was wanting to be able to send calendar invites for personal events, e.g. holiday flights.

I've managed to use sendmail as included by the experimental images but that required a number of hacks:

* One obviously needs to start the sendmail daemon.

* The first problem one might encounter is that sendmail needs a FQDN in the hosts file, otherwise it hangs for a minute every time a mail is sent. (and the daemon is started). One needs to add to the hosts file. You can't even use `sed -i` directly due to the hosts file being mounted (and not moveable).
* Your mail wil literally not be accepted by anyone unless you have a proper mail server e.g. postfix somewhere on the network. You can do this by adding a SMART_HOST to sendmail.mc
* Even then, your emails may still go to spam, because the `mail()` function as used by postfix doesn't set the envelope sender (via the  `-f`flag)

This is the command I'm using that actually delivers mail properly to gmail and the like:
```
              mv /usr/sbin/sendmail /usr/sbin/sendmail-og && 
              echo '#!/bin/sh' > /usr/sbin/sendmail && chmod +x /usr/sbin/sendmail &&  
              echo '/usr/sbin/sendmail-og -f baikal@yourlocal.domain $@' >> /usr/sbin/sendmail &&
              echo 'nginx' >> /etc/mail/trusted-users &&
              echo "define(\`SMART_HOST', \`mail.yourlocal.domain')dnl" >> /etc/mail/sendmail.mc &&
              cp /etc/hosts /etc/hosts.new && 
              sed -i "s/$HOSTNAME/$HOSTNAME baikal.yourlocal.domain/" /etc/hosts.new && 
              cp /etc/hosts.new /etc/hosts && 
              service sendmail start && 
              /etc/init.d/php8.0-fpm start && 
              chown -R nginx:nginx /var/www/baikal/Specific && 
              nginx -g "daemon off;"
```

If any one is interested in the k8s manifest I used:
https://github.com/vaskozl/home-infra/commit/9bbdf3d611085edff0309fc0fcfe63cda2e73f1e

**Comment by ckulka** [comment]
Thanks all, especially @philippneugebauer and @vaskozl. My proposal below is basically just a mix of #32 and https://github.com/ckulka/baikal-docker/issues/31#issuecomment-1008344665.

Would like to hear what you think of this:

1. The Docker image won't contain a mail server, it will rely on an external one (another container or a service like Gmail)

2. The container will use sendmail and not msmtp

   Seems like sendmail is more widely adopted (<https://trends.google.de/trends/explore?q=sendmail,msmtp>), looks like being more of the default option to go with, and judging from the releases both are about equally frequently maintained.

   It also seems easier to use mounted secrets with sendmail compared to msmtp.

3. Sendmail configuration will be applied during container start using three environment variables: the host, the "from" email address, and the additional `sendmail.mc` configuration

The `docker-compose.yaml` file would look like this:

```yaml
version: '2'
services:
  baikal:
    image: ckulka/baikal:nginx
    environment:
      BAIKAL_SERVERNAME: baikal.yourdomain.com
      SENDMAIL_FROM: baikal@yourdomain.com
      SENDMAIL_MAILER_DEFINITIONS: |
        define(\`SMART_HOST', \`mail.yourdomain.com')dnl
```

The `SENDMAIL_FROM` environment variable with a prefixed `-f` flag is added to `mail.force_extra_parameters` in `/etc/php/8.1/fpm/php.ini`, so that we don't have to customize the `/usr/sbin/sendmail` script.

The `SENDMAIL_MAILER_DEFINITIONS` environment variable is inserted into `/etc/mail/sendmail.mc` underneath `MAILER_DEFINITIONS`, according to [Configure Sendmail](https://linuxconfig.org/configuring-gmail-as-sendmail-email-relay) it's important that it's right underneath it ¯\ _ (ツ)_/¯. It can have multiple lines to allow for more configuration like authentication.

The `BAIKAL_SERVERNAME` is inserted to `/etc/hosts` just like @vaskozl has it in the script. **What I don't know/my question:** is this the same as the FQDN where Baikal is accessed, or could this be a different one?

If it's always the same one, we can reuse `BAIKAL_SERVERNAME` which is already a (somewhat) required parameter for the Apache image. If it can be different, I'd opt for a different variable name.

**Reply by philippneugebauer** [reply]
I guess 1 is a good idea. You rather want a separate mailserver but probably even a hosted one.
2: I just went for the easiest solution to implement and it seemed to me it's msmtp, so I have no opinion on that

**Reply by philippneugebauer** [reply]
3 sounds reasonable

**Comment by vaskozl** [comment]
> What I don't know/my question: is this the same as the FQDN where Baikal is accessed, or could this be a different one?

No. It doesn't even need to be a valid FQDN. But it will be used to determine the envelope sender (e.g. nginx@$fqdn) in the case `-f` isn't  specified and is also used in the outgoing SMTP HELO. Some mailservers may not be happy if you send in a HELO with a FQDN which doesn't resolve, while others may be happy with a basic (technically incorrect) `HELO localhost`.

I personally use `msmtp` on my desktop to interact with external email providers, like gmail. It has a few advantages in my mind:
* It doesn't require a sendmail service/daemon to be running
* The configs are simpler than sendmail
* It's easier to configure authentication SMTP (if you dont' have an open local relay handy)

I don't really have a preference and would be happy with either at the end of the day.

**Comment by ckulka** [comment]
Regarding sendmail vs. msmtp, that's fair, I have no experience with either and just went off of comparing online presence on Google search results and Stackoverflow - but you could make an argument that sendmail has more because it's not as intuitive and easy to set up. Someone on reddit ranked it as "archaic where nobody knows how the configuration works" :)

I also found out that the password secrets aren't hard either, because the environment variable with the configuration and password can (for example in Kubernetes) be stored in a secret, so no problem there - and even easier then in msmtp!

I'll push an update to the [features/mail](https://github.com/ckulka/baikal-docker/tree/features/mail) branch with the configuration example from #32, does someone know/have a configuration for a simple SMTP server without security settings like auth or TLS?

**Comment by ckulka** [comment]
I updated the feature branch, could you please review [docs/email-guide.md](https://github.com/ckulka/baikal-docker/blob/features/mail/docs/email-guide.md)?

If the guide and the PR #63 in general looks good, I'll merge into `master` and we can test the experimental images (which btw is now on Baikal 0.9.1). If you have time to build the feature branch locally and test ahead of time... even better :)

@philippneugebauer I wanted to continue from your PR, but couldn't because of a merge conflict. I don't necessarily want to loose your commit/contribution from the Git history, if you rebase on the current `master` branch I'm happy to include your commits in the feature branch. Just let me know.

**Reply by philippneugebauer** [reply]
Honestly, I appreciate your effort, but I don't care :)

**Comment by ahgraber** [comment]
I've build a [version of the `nginx` image](https://hub.docker.com/r/ahgraber/baikal/tags) based on `features/mail` with one change --  I had to move the `apt install msmtp msmtp-mta` down to the final build because the binaries weren't being copied from the builder image.  

However, after doing so, I can confirm that the images work and do send email invites.  I'm [running baikal as a k3s deployment](https://github.com/ahgraber/homelab-gitops-k3s/tree/main/cluster/apps/services/baikal), so I took inspiration from docs/email-guide.md, but had some additional tweaks required to get it working as a deployment while mounting `msmtprc` as a secret, etc.

If I can figure out how to do so, I'll set a PR against `features/mail`.

**Comment by raenrfm** [comment]
I've set up Baikal in a docker container on a hosted server, I have smtp enabled from the host provider but sendmail doesn't work with the latest nginx build I also have an email server set up in an additional container and it is working with emails, it's just calendar invites that do not.  Is there something other than defining the email address that the invites should come from?  Do I need to enable additional env variables per above or has this already been rolled into the latest builds?

**Comment by ckulka** [comment]
Finally! I got around cleaning up and publishing the documentation, examples, and updated Docker images for msmtprc:

* [docs/email-guide.md](https://github.com/ckulka/baikal-docker/blob/master/docs/email-guide.md)
* `ckulka/baikal:experimental-apache`
* `ckulka/baikal:experimental-nginx`

Could you please give it a try and let me know if things work as expected? If so, I'll see that I release this before the next Baikal release, probably as something like `ckulka/baikal:0.9.3-msmtprc` (and the other Apache httpd and nginx variant tags).

**Reply by benlenau** [reply]
Works! Looking forward to the next release :-)

**Comment by ahgraber** [comment]
I'm no longer relying on baikal for my dav requirements, but I spun up a test container using `experimental-nginx` and everything seemed to work as expected.  This was a quick test, so I can't validate over a longer period.

**Comment by MrPinrel** [comment]
I also moved from baikal to nextcloud so I’m sorry I can’t help…..Marcos*Sent from my iPhoneOn Dec 31, 2022, at 16:25, Alex Graber ***@***.***> wrote:﻿
I'm no longer relying on baikal for my dav requirements, but I spun up a test container using experimental-nginx and everything seemed to work as expected.  This was a quick test, so I can't validate over a longer period.

**Comment by vaskozl** [comment]
Thanks for this. I tested `ckulka/baikal:experimental-nginx@sha256:ea235c8de2af002ca44a9842fb55f6c25e676266cd358c2af3
8a1ab28672e74b` and it appears to work perfectly via a php shell and when using sendmail directly.

Unfortunately I've also just discovered that IOS 16.3 may be no longer compatible as I was having some trouble sending it out via the client.

**Reply by ckulka** [reply]
Do you mean the email-aspect of it, or that Baikal generally doesn't work with iOS 16.3?

Asking because the latter might be something that [sabre-io/Baikal](https://github.com/sabre-io/Baikal) has to fix, but emailing something that I'd have to look into.

**Reply by benlenau** [reply]
Invites (e-mail) work fine for me on iOS 16.3 (using builtin iOS calendar) and _ckulka/baikal:experimental-apache image id 4916d0cf2eb2_

**Reply by vaskozl** [reply]
Sorry, I'm an idiot and forgot to click upgrade to 9.3 from 9.2 on the web interface.

It works with the ios calendar perfectly!

**Comment by vaskozl** [comment]
Cheers for this. I definitely think msmtp is better, especially in a container without a mail daemon.

Out of interest, do you guys know of an Android calendar client which works well with this feature, my current phone is an iPhone and the stock calendar app works great, but might not be my next phone and isn't for everyone in my household.

**Reply by DelusionalLogic** [reply]
The stock samsung calendar app and Davx5 works well.

**Comment by DelusionalLogic** [comment]
I've tested nginx-experimental, and it seems to work! I added a `logfile` directive to the msmtprc, but otherwise is just worked out of the box.

**Comment by tedstriker** [comment]
I'm about to install Baïkal for it's ability to send out invitations. 
But I wonder how replies (will attend/decline) will be handled. Is this handling available as well or does it only send .ics-files containing the appointment?

**Comment by ckulka** [comment]
I released [0.9.3+msmtp](https://github.com/ckulka/baikal-docker/releases/tag/0.9.3%2Bmsmtp) and the new images have been pushed to Docker Hub, so the 0.9.3 images now include msmtp.

#131 reminded me that it's been a while and since there weren't any issues, it was high time to publish it!

**Reply by Judman** [reply]
@ckulka I might be dense, but I cannot get an invite to send via email. I have test msmtp and it will send an email out through my hosted mailbox, but when I add a calendar event with an added participant the invite never arrives. Is there a Baikal or sabre/dav log location I should be checking. Sorry for the entry level questions. I know just enough to get myself in trouble.

Below are my configs with sensitive details replaced with example data.

cat /etc/msmtprc 
```
defaults
auth           on
tls            on
tls_trust_file /etc/ssl/certs/ca-certificates.crt
account        default
host           hosted.example.org
port           587
from           calendar@example.org
user           calendar@example.org
password       **password**
```
echo "This is a test e-mail from my server using msmtp!" | msmtp email@example.com

Result:
Nov 07 04:18:46 host=hosted.example.org tls=on auth=on user=calendar@example.org from=calendar@example.org recipients=email@example.com mailsize=181 smtpstatus=250 smtpmsg='250 2.0.0 Ok: queued as DF8845C04F0' exitcode=EX_OK

**Reply by MrPinrel** [reply]
Did you ever get it to work or find out how to troubleshoot? I am having the same issue. Nothing is sent and I don't really know how to troubleshoot. I am using docker inside a Synology box.

**Reply by PintjesB** [reply]
I am stuck at this as well, can not figure it out...

**Reply by thomasjanker** [reply]
For me it currently works well with nginx variant.

Some time ago I had also troubles with sending mails. Was a permission issue, but is solved now. 

Debugging mail delivery issues is not easy with msmtp, due to missing logs. 

Try to send mail from within the container, to determine whether it is a config issue or not.

**Reply by PintjesB** [reply]
@thomasjanker Did you switch to the nginx variant in order to get it working or just because...?

Also, do you remember what file needed permissions? I read somewhere on dietpi forums that /etc/msmtprc needed 0600 perms, have tried that and sadly didn't change anything...

**Reply by thomasjanker** [reply]
I'm using nginx because of memory footprint. 

Exactly, /etc/msmtprc permission caused the troubles. Switching to 0666 was needed to allow access for uid nginx is running.

**Reply by PintjesB** [reply]
Sadly that did not seem to fix it... Like you said very hard to troubleshoot as it logs nothing...

**Reply by since9629** [reply]
Same issue for me as well, running baikal inside docker container, @Judman were you able to find a fix for this ?
